lib/xdrgen/generators/go.rb

```diff
@@ -549,6 +531,259 @@ def render_encode_to_body(out, var, type, self_encode:)
+      # xdr.Decoder, and a variable defined by `var` that is the value to
+      # encode.
+      def render_decode_from_body(out, var, type, declared_variables:, self_encode:)
+        tail = <<-EOS.strip_heredoc
+          n += nTmp
+          if err != nil {
+            return n, err
+          }
+        EOS
+        optional = type.sub_type == :optional
+        if optional
+          render_variable_declaration(out, "  ", 'b', "bool", declared_variables: declared_variables)
+          out.puts "  b, nTmp, err = d.DecodeBool()"
+          out.puts tail
+          out.puts "  #{var} = nil"
+          out.puts "  if b {"
+          out.puts "     #{var} = new(#{name type})"
+        end
+        case type
+        when AST::Typespecs::UnsignedHyper
+          out.puts "  #{var}, nTmp, err = d.DecodeUhyper()"
+          out.puts tail
+        when AST::Typespecs::Hyper
+          out.puts "  #{var}, nTmp, err = d.DecodeHyper()"
+          out.puts tail
+        when AST::Typespecs::UnsignedInt
+          out.puts "  #{var}, nTmp, err = d.DecodeUint()"
+          out.puts tail
+        when AST::Typespecs::Int
+          out.puts "  #{var}, nTmp, err = d.DecodeInt()"
+          out.puts tail
+        when AST::Typespecs::String
+          arg = "0"
+          arg = type.decl.resolved_size if !type.decl.resolved_size.nil?
+          out.puts "  #{var}, nTmp, err = d.DecodeString(#{arg})"
+          out.puts tail
+        when AST::Typespecs::Opaque
+          if type.fixed?
+            out.puts "  nTmp, err = d.DecodeFixedOpaqueInplace(#{var}[:])"
+            out.puts tail
+          else
+            arg = "0"
+            arg = type.decl.resolved_size if !type.decl.resolved_size.nil?
+            out.puts "  #{var}, nTmp, err = d.DecodeOpaque(#{arg})"
+            out.puts tail
+          end
+        when AST::Typespecs::Simple
+          case type.sub_type
+          when :simple, :optional
+            optional_within = type.is_a?(AST::Identifier) && type.resolved_type.sub_type == :optional
+            if optional_within
+              render_variable_declaration(out, "  ", 'b', "bool", declared_variables: declared_variables)
+              out.puts "  b, nTmp, err = d.DecodeBool()"
+              out.puts tail
+              out.puts "  #{var} = nil"
+              out.puts "  if b {"
+              out.puts "     #{var} = new(#{name type.resolved_type.declaration.type})"
+            end
+            var = "(*#{name type})(#{var})" if self_encode
+            out.puts "  nTmp, err = #{var}.DecodeFrom(d)"
+            out.puts tail
+            if optional_within
+              out.puts "  }"
+            end
+          when :array
+            out.puts "  for i := 0; i < len(#{var}); i++ {"
+            element_var = "#{var}[i]"
+            optional_within = type.is_a?(AST::Identifier) && type.resolved_type.sub_type == :optional
+            if optional_within
```

Comment:
Intuitively, it seems like this optional block (and above) should be able to be handled with a recursive data-type, like a `Option<Foo>` type-thing, and having a `render_optional` method... 🤔 but this is probably good enough for now.

Reply:
I think I'm thinking something like:
```ruby

def render_decode_from_body(out, var, type, declared_variables:, self_encode:)
  # ...
  if type.sub_type == :optional
    return render_optional(out, var, type, type, declared_variables: declared_variables, self_encode: self_encode) do |inner_type|
      render_decode_from_body(out, var, inner_type, declared_variables: declared_variables, self_encode: self_encode)
    end
  end
  # ...
end

def render_optional(out, var, type, declared_variables:, self_encode:)
  render_variable_declaration(out, "  ", 'b', "bool", declared_variables: declared_variables)
  out.puts "  b, nTmp, err = d.DecodeBool()"
  out.puts tail
  out.puts "  #{var} = nil"
  out.puts "  if b {"
  out.puts "     #{var} = new(#{name type.resolved_type.declaration.type})"
  # TODO: Unwrap the inner_type (without the optional wrapper) here
  yield inner_type
  out.puts "  }"
end
```

But, actually you could do a similar block thing for arrays as well, I think...

Reply:
Either way, probably outside the scope of this PR.